Refactor and version 0.0.2

JensRantil opened the pull request:
Major changes are:

  * Make collection synchronous. This is best practise according to [1]
    and fixes [2]. This fixes #6.
  * Dropped the harness. Felt a "Prometheus framework" was unnecessary.
  * Restructured the code according to [3].
  * Improved testing.
    * Added an `interface` for the implementation that queries Kafka to
      be able to test it more thoroughly.
    * All tests pass.
    * Test coverage is now 72.45%.
  * Added support for querying partitions concurrently.
    * Added concurrency limit on the number pending group queries to avoid
      starving resources if there are many groups. Haven't tested this in
      production, though, as we only have a few consumer groups.
  * Improved logging to simplify debugging.
  * Improved error handling when the exporter can't connect to Kafka.
  * Instrument scraping errors. This fixes #3.
  * Adding timestamps to Kafka metrics to accomodate for slow scraping for high
    throughput/infrequent committing consumer groups. This is a case that was raised by
    @kawamuray which is a case I have not personally experienced.

[1] https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling
[2] https://github.com/kawamuray/prometheus-kafka-consumer-group-exporter/issues/6
[3] https://medium.com/@benbjohnson/standard-package-layout-7cdbc8391fc1#.kpbltioae

JensRantil committed:
refactor(packaging): remove unnecessary package

Cleaner that way. This change means that `go build .` will build a
binary called `prometheus-kafka-consumer-group-exporter` instead of
`kafka-consumer-group-exporter`.

JensRantil committed:
refactor: default to non-public constants/variables

While this doesn't matter for `main` packages, I consider it a good
practise in case moving stuff out to separate packages.

JensRantil committed:
feat: major overhaul

 * Make collection synchronous. This is best practise according to [1]
   and fixes [2].
 * Dropped the harness. Felt a "Prometheus framework" was unnecessary.
 * Restructured the code according to [3].
 * Improved testing.
   * Added an `interface` for the implementation that queries Kafka to
     be able to test it more thoroughly.
   * All tests pass.
 * Added support for querying partitions concurrently.
   * Added concurrency limit on the number pending group queries.

[1] https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling
[2] https://github.com/kawamuray/prometheus-kafka-consumer-group-exporter/issues/6
[3] https://medium.com/@benbjohnson/standard-package-layout-7cdbc8391fc1#.kpbltioae

JensRantil committed:
chore: new version 0.0.2

JensRantil committed:
log: include broken line on error

Will help in debugging.

JensRantil committed:
fix: register connection issues with Kafka as errors

The script will still exit with exit code 0. Must be determined through
output.

JensRantil committed:
refactor(cmd): use `logrus` for logging

JensRantil committed:
log: log why listing groups failed

JensRantil committed:
refactor(kafka): avoid one level of indirection

...by constructing one-off `Metric`s for Kafka metrics. This seem to be
best-practise according to [1].

[1] https://godoc.org/github.com/prometheus/client_golang/prometheus

JensRantil committed:
fix(prometheus): add timestamp to Kafka metrics

See inline patch comment for the reasoning behind this.

JensRantil committed:
fix(cmd): check error on CLI run

Found by `errcheck`.

JensRantil commented:
This is reviewable. Remaining:

 - [x] Document all public variables, methods.
 - [x] Run `go vet´.
 - [x] Run `errcheck`.
 - [x] Run `golint`.
 - [x] Run `unused`.
 - [x] Run `gocyclo`.

JensRantil committed:
refactor: use field names for struct

Proposed by `golint`.

JensRantil committed:
refactor: reduce cyclomatic complexity a bit

Proposed by `gocyclo`.

JensRantil commented:
Also, naming of methods and variables can probably be improved (apart from what `golint` proposed, that is).

JensRantil commented:
I'll run the last mile with this pull request tomorrow.

JensRantil committed:
refactor(cmd): don't export version

No reason to do so.

JensRantil committed:
refactor(prom): make TimestampedMetric private

No reason to expose it outside of the package.

JensRantil committed:
doc: document all public types and functions

`golint` helped out here.

JensRantil committed:
refactor: PartitionId => PartitionID

Golang naming best-practise.

JensRantil committed:
refactor: ClientId => ClientID

Golang naming best-practise.

JensRantil committed:
refactor: clientIdAndConsumerAddress => clientIDAndConsumerAddress

Golang naming best-practise.

JensRantil committed:
refactor: clientId => clientID

Golang naming best-practise.

JensRantil committed:
fix(main): don't log `nil` on `--help`

JensRantil committed:
ref: only single bootstrap server taken by script

Command line arguments unaltered.

JensRantil commented:
I now consider this PR fully ready for review.

JensRantil committed:
fix: avoid overlapping scrapes

We've seen many `kafka-consumer-groups.sh` processes spawned from the
exporter. This will make a new scrape reuse the result from a previous
scrape if it's already running.

JensRantil committed:
fix(sync): avoid race condition

...which I couldn't find by rewriting the logic to be channel based
instead.

JensRantil committed:
doc(sync): document public functions

JensRantil commented on kafka/collector.go:
```diff
@@ -0,0 +1,133 @@
+func (col *ConsumerGroupClient) execConsumerGroupCommand(args ...string) (string, error) {
+	allArgs := append([]string{"--new-consumer", "--bootstrap-server", col.BootstrapServer}, args...)
+	cmd := exec.Command(col.ConsumerGroupCommandPath, allArgs...)
+	output, err := cmd.CombinedOutput()
```
We've seen `--list` script execution lock-up forever. No sure why. Before merging this pull request, we should probably add a timeout here that kills that script after a while.

JensRantil replied:
Fixed in a9f87ca0b1906d15c36df612d9ee96ec01cdb0b7.

JensRantil committed:
fix(kafka): kill the script if it runs more than timeout

We've seen this happen in production 2-3 times. We suspect this happens
during rebalances, but we don't know strictly.

JensRantil committed:
ref(scraping): remove unnecessary semaphore

We don't need this since 27f110b limits the number of concurrent calls
to the Kafka script, which was what this prohibited.

ashoksahoo commented:
When can we expect this to be merged?

kawamuray commented on prometheus/collector.go:
```diff
@@ -0,0 +1,156 @@
+		}),
+		prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "kafka_broker_consumer_group_describe_errors",
+			Help: "Number of Kafka scraping errors.",
```
Need to differentiate help string to above one?(don't remember well but IIUC prometheus warns when it founds duplicated help messages?)

JensRantil replied:
That's a good question and a valid point! I suggest we push that forward in the future and fix that in a separate pull request (if needed). If you want, we can create an issue for it later.

kawamuray replied:
sounds good :)

kawamuray commented on README.md:
```diff
@@ -36,5 +36,5 @@ Example Usage
 # Download latest Kafka distribution(if necessary)
 tar zxvf kafka_LATEST_VERSION.tgz
-./kafka_consumer_group_exporter --consumer-group-command-path=./kafka_LATEST_VERSION/bin/kafka-consumer-groups.sh BOOTSTRAP_SERVERS
+./kafka_consumer_group_exporter --consumer-group-command-path=./kafka_LATEST_VERSION/bin/kafka-consumer-groups.sh BOOTSTRAP_SERVER
```
Any intentions to make this singular form over plural form?
I know somehow the help message of `kafka-consumer-groups.sh` uses singular form as an placeholder of `bootstrap.servers`, but I think it's always preferred to pass multiple bootstrap servers for this exporter argument to make it keep working even when one broker went down.

JensRantil replied:
I changed to singular because I thought the help text for `kafka-consumer-groups.sh` hints that it only takes a single server:
```sh
$ /Users/jens/Development/src/external/kafka_2.11-0.10.0.0/bin/kafka-consumer-groups.sh
List all consumer groups, describe a consumer group, or delete consumer group info.
Option                                 Description
------                                 -----------
--bootstrap-server <server to connect  REQUIRED (only when using new-
  to>                                    consumer): The server to connect to.
--command-config <command config       Property file containing configs to be
  property file>                         passed to Admin Client and Consumer.
--delete                               Pass in groups to delete topic
                                         partition offsets and ownership
                                         information over the entire consumer
                                         group. For instance --group g1 --
                                         group g2
                                       Pass in groups with a single topic to
                                         just delete the given topic's
                                         partition offsets and ownership
                                         information for the given consumer
                                         groups. For instance --group g1 --
                                         group g2 --topic t1
                                       Pass in just a topic to delete the
                                         given topic's partition offsets and
                                         ownership information for every
                                         consumer group. For instance --topic
                                         t1
                                       WARNING: Group deletion only works for
                                         old ZK-based consumer groups, and
                                         one has to use it carefully to only
                                         delete groups that are not active.
--describe                             Describe consumer group and list
                                         offset lag related to given group.
--group <consumer group>               The consumer group we wish to act on.
--list                                 List all consumer groups.
--new-consumer                         Use new consumer.
--topic <topic>                        The topic whose consumer group
                                         information should be deleted.
--zookeeper <urls>                     REQUIRED (unless new-consumer is
                                         used): The connection string for the
                                         zookeeper connection in the form
                                         host:port. Multiple URLS can be
                                         given to allow fail-over.
```
Am I wrong? Does the help text differ for later Kafka versions, perhaps?

kawamuray replied:
AFAIK, it takes multiple servers list. The value given by that option will eventually passed to `bootstrap.servers` configuration of `AdminClient`, which is capable to accept multiple servers: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala#L452

I'm not sure why Kafka's help text is in singular form, but I'd think it's better to pass multiple servers for this exporter because of this reason I explained previously:
> preferred to pass multiple bootstrap servers for this exporter argument to make it keep working even when one broker went down.

kawamuray commented on prometheus/collector.go:
```diff
@@ -0,0 +1,156 @@
+	var wg sync.WaitGroup
+	wg.Add(len(groupnames))
+	for _, groupname := range groupnames {
+		go func(groupname string) {
```
What do you think of limiting the maximum concurrency applied here for querying each group's description in parallel?
I'm not fully aware how expensive is it to ask brokers for describing a group, but considering the response time we've experienced it doesn't sounds a lightweight operation.
The current implementation runs queries of the number in propotion to the number of groups, which can be very huge in some big clusters I think.
As an another thought, the running server's resource maybe exhausted due to too many java processes launched at once?

JensRantil replied:
Good idea. We currently only have a single topic and will soon be adding a second one. I added a default max concurrency limit of 20 in ab8217ed7eedd28d39e511134328401295ad7b06. Do you think that's a sane limit?

kawamuray commented on prometheus/collector.go:
```diff
@@ -0,0 +1,156 @@
+			cancel()
+			if err != nil {
+				p.groupDescribeErrors.WithLabelValues(groupname).Inc()
+				return
```
How about logging at least the name of group which has failed to get its detail? Would be worth to figure out what the problem was later on.

JensRantil replied:
Good idea. Fixed in 9b143de00e0389d5d7fd3fab80c8df82b2ff81f0.

kawamuray commented on metrics.go:
```diff
@@ -0,0 +1,20 @@
+
+// PartitionInfoClient queries consumer groups and consumer group partitions
+// for their stats.
+type PartitionInfoClient interface {
```
For namings would it be more intuitive to fix names like below? WDYT?

- `PartitionInfoClient` => `ConsumerGroupInfoClient`
- `ConsumerGroupClient` => `ConsumerGroupsCommandClient`
- `FanInPartitionInfoClient` => `FanInConsumerGroupInfoClient`

JensRantil replied:
Good idea. Fixed in 65e9e49 & 154a157.

kawamuray reviewed (changes requested):
@JensRantil I'm sorry for being late to take a look at this. I learned a lot from your highly sophisticated Go code through reviewing this PR :)

Just left few comments. PTAL.

kawamuray commented:
@ashoksahoo Sorry for taking a long time to work on this. I expect this to be merged soon.

JensRantil committed:
ref(naming): PartitionInfoClient => ConsumerGroupInfoClient

JensRantil committed:
ref(naming): ConsumerGroupClient => ConsumerGroupsCommandClient

JensRantil committed:
log(collector): log when unable to query group

Cred to Yuto for pointing this out.

JensRantil committed:
fea(collector): set max concurrent group queries

Defaults to 20. Cred to Yuto for pointing this out.

JensRantil committed:
Merge remote-tracking branch 'upstream/master' into on-demand-scraping

* upstream/master:
  README improvements (#7)

JensRantil commented:
@kawamuray Care to have a look at the latest commits I just pushed?

By the way, we've been running 36a9342bd10d34bc8d401d0e612f819d7ffeeac4 in production for some time now and haven't had any issues.

JensRantil commented:
...also let me know if you'd like me to rebase on top of `master` to get a cleaner git history.

JensRantil committed:
fix(collector): avoid deadlock

Did a last minute chang before pushing to Github which lead to a
deadlock...

kawamuray commented on cmd/kafka_consumer_group_exporter.go:
```diff
@@ -0,0 +1,77 @@
+		cli.IntFlag{
+			Name:  "max-concurrent-group-queries",
+			Usage: "The maximum number of consumer groups that are queried concurrently.",
+			Value: 20,
```
> Do you think that's a sane limit?

https://github.com/apache/kafka/blob/trunk/bin/kafka-run-class.sh#L210
Given that Kafka defaults maximum heap size to 256M for commands, `20` sounds little dangerous depending on where users runs this exporter.
I don't think consumer group command really allocates 256MB on each run but potentially it means this exporter can allocate `256M * 20 = 5GB` at maximum just for reporting metrics, which might lead memory contention with other processes that are more important.
The worst case I imagine here is that some users run this exporter on a server which also runs application processes and this exporter disturbs its memory allocation periodically ;;
We need to take a balance between safety and better usability, so I'd suggest to set this default to `4`, which potentially allocates memory up to 1G+ bytes by default.
WDYT?

JensRantil replied:
Very good point! Corrected in 4311a211a091aa4c66e29364773820c1b2593b44. Also added a code comment about the reasoning behind the default value just in case.

kawamuray commented on prometheus/collector.go:
```diff
@@ -0,0 +1,172 @@
+			if err != nil {
+				log.Error("Could not describe group '%s': %s", groupname, err)
+				p.groupDescribeErrors.WithLabelValues(groupname).Inc()
+				return
```
shouldn't this be changed to `continue` to keep processing remaining groupnames?

JensRantil replied:
Absolutely. Nice find. Fixed in a7910be748796a9291b16f28dc0b911bd2dcc8cd.

kawamuray commented on README.md:
```diff
@@ -42,5 +42,5 @@ Example Usage
 # Download latest Kafka distribution(if necessary)
 $ tar zxvf kafka_LATEST_VERSION.tgz
-$ ./kafka_consumer_group_exporter --consumer-group-command-path=./kafka_LATEST_VERSION/bin/kafka-consumer-groups.sh BOOTSTRAP_SERVERS
+$ ./kafka_consumer_group_exporter --consumer-group-command-path=./kafka_LATEST_VERSION/bin/kafka-consumer-groups.sh BOOTSTRAP_SERVER
```
> I changed to singular because I thought the help text for kafka-consumer-groups.sh hints that it only takes a single server: 

AFAIK, it takes multiple servers list. The value given by that option will eventually passed to bootstrap.servers configuration of AdminClient, which is capable to accept multiple servers: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala#L452

I'm not sure why Kafka's help text is in singular form, but I'd think it's better to pass multiple servers for this exporter because of this reason I explained previously:

> preferred to pass multiple bootstrap servers for this exporter argument to make it keep working even when one broker went down.

JensRantil replied:
Interesting. I didn't dig into this too much, but I assume it's comma separated? 1a0f1df3bb995d1442d551690742ce9de4ce4855 looks good?

kawamuray replied:
@JensRantil  yep, looks good. and this readme text as well?

JensRantil replied:
Whoops. Force-pushed a fix for that.

JensRantil replied:
Oh, I wanted to say here that I've never had any reason to use multiple servers here. The way we do scraping is that we are running the exporter on every Kafka broker. Each exporter is only connecting to localhost. That way, we can use the exporter to see if Kafka down on a specific broker while still getting HA since we are scraping all Kafka brokers. Your mileage might vary, but I thought this could be interesting for you to hear/know.

kawamuray reviewed (changes requested):
Thanks for update @JensRantil !

Left few more comments.

kawamuray commented:
> ...also let me know if you'd like me to rebase on top of master to get a cleaner git history.

I'm going to use "Squash and merge" button to merge this PR so you can leave current history as-is.
If you prefer to organize commits by yourself, it's also fine for me :)

JensRantil committed:
ref(max-concurrency): default to 4 instead of 20

See [1].

[1] https://github.com/kawamuray/prometheus-kafka-consumer-group-exporter/pull/9#discussion_r116469847

JensRantil committed:
fix(collector): continue on error

Nice find by Yuto.

JensRantil committed:
ref(collector): multiple bootstrap servers supported

JensRantil commented:
> I'm going to use "Squash and merge" button to merge this PR so you can leave current history as-is.
> If you prefer to organize commits by yourself, it's also fine for me :)

Nah, "Squash and merge" sounds good to me.

JensRantil commented:
@kawamuray Feel free to do another review. Looks better?

kawamuray commented:
LGTM 👍 

@JensRantil Thanks again for your great contribution. I'll test this in our environment when I could find time :)

@ashoksahoo FYI

ashoksahoo commented:
Thank you!
